crates/codegen/parser/src/code_generator.rs

```diff
@@ -3,47 +3,121 @@ use std::{
     path::PathBuf,
 };
 
+use codegen_schema::types::grammar::Grammar;
 use codegen_utils::context::CodegenContext;
 use inflector::Inflector;
 use proc_macro2::{Ident, TokenStream};
 use quote::{format_ident, quote};
 use semver::Version;
 
-#[derive(Clone, Debug, Default)]
-pub struct VersionedFunctionBody {
-    pub comment: String,
-    pub versions: BTreeMap<Version, TokenStream>,
-}
+#[derive(Clone, Debug)]
+pub struct VersionedFunctionBody(BTreeMap<Version, Option<(String, TokenStream)>>);
 
 impl VersionedFunctionBody {
-    fn insert(&mut self, comment: &str, version: &Version, body: TokenStream) {
-        self.comment.push_str(comment);
-        self.versions.insert(version.clone(), body);
+    fn new(version: &Version) -> Self {
+        let mut versions = BTreeMap::new();
+        versions.insert(version.clone(), None);
+        Self(versions)
     }
 
-    fn to_function_body(&self) -> TokenStream {
-        let first_version = self.versions.iter().next().unwrap().0;
-        self.versions
-            .iter()
-            .rev()
-            .map(|(version, body)| {
-                let version_flag = format_ident!(
-                    "version_is_equal_to_or_greater_than_{version}",
-                    version = version.to_string().replace(".", "_")
-                );
-                if version == first_version {
-                    quote! { { #body } }
-                } else {
-                    quote! { if self.#version_flag { #body } }
-                }
+    fn insert(&mut self, version: &Version, definition: Option<(String, TokenStream)>) {
+        self.0.insert(version.clone(), definition);
+    }
+
+    fn is_defined_for_all_versions(&self) -> bool {
+        !self.0.values().any(|f| f.is_none())
```

Comment:
nit: `all()` vs `!any()`?

Also, not sure if this should check that the first defined version is the first version of the grammar? in case a production was introduced later.

```suggestion
        self.0.values().all(|f| f.is_some())
```

Reply:
done